### ISSUE TITLE
[Wasm] illegal cast exception when trying to use Firebase Auth with Flutter

### ISSUE THREAD

**Issue by sma** [dart-lang/sdk#55513]
In a Flutter web project that includes (and uses, but not even calls) Firebase Auth, I get this exception:

```
TrustedTypes available. Creating policy: flutterfire-firebase_core
VM216:2 Initializing Firebase firebase_core
main.dart.mjs:587 TrustedTypes available. Creating policy: flutterfire-firebase_firestore
VM217:2 Initializing Firebase firebase_firestore
main.dart.mjs:587 TrustedTypes available. Creating policy: flutterfire-firebase_auth
VM218:2 Initializing Firebase firebase_auth
main.dart.wasm:0x1505e7 Uncaught RuntimeError: illegal cast
    at Auth.onWaitInitState inner (main.dart.wasm:0x1505e7)
    at _awaitHelperWithTypeCheck closure at org-dartlang-sdk:///dart-sdk/lib/_internal/wasm/lib/async_patch.dart:97:16 (main.dart.wasm:0x14fd65)
    at closure wrapper at org-dartlang-sdk:///dart-sdk/lib/_internal/wasm/lib/async_patch.dart:97:16 trampoline (main.dart.wasm:0x14fdcc)
    at _RootZone.runUnary (main.dart.wasm:0x13b242)
    at _Future._propagateToListeners (main.dart.wasm:0x13adeb)
    at _Future._completeWithValue (main.dart.wasm:0x13aa06)
    at _Future._asyncCompleteWithValue closure at org-dartlang-sdk:///dart-sdk/lib/async/future_impl.dart:721:29 (main.dart.wasm:0x13a9b3)
    at closure wrapper at org-dartlang-sdk:///dart-sdk/lib/async/future_impl.dart:721:29 trampoline (main.dart.wasm:0x13a9dd)
    at _startMicrotaskLoop (main.dart.wasm:0x13b6f9)
    at _startMicrotaskLoop tear-off trampoline (main.dart.wasm:0x13b763)
$Auth.onWaitInitState inner @ main.dart.wasm:0x1505e7
$_awaitHelperWithTypeCheck closure at org-dartlang-sdk:///dart-sdk/lib/_internal/wasm/lib/async_patch.dart:97:16 @ main.dart.wasm:0x14fd65
$closure wrapper at org-dartlang-sdk:///dart-sdk/lib/_internal/wasm/lib/async_patch.dart:97:16 trampoline @ main.dart.wasm:0x14fdcc
$_RootZone.runUnary @ main.dart.wasm:0x13b242
$_Future._propagateToListeners @ main.dart.wasm:0x13adeb
$_Future._completeWithValue @ main.dart.wasm:0x13aa06
$_Future._asyncCompleteWithValue closure at org-dartlang-sdk:///dart-sdk/lib/async/future_impl.dart:721:29 @ main.dart.wasm:0x13a9b3
$closure wrapper at org-dartlang-sdk:///dart-sdk/lib/async/future_impl.dart:721:29 trampoline @ main.dart.wasm:0x13a9dd
$_startMicrotaskLoop @ main.dart.wasm:0x13b6f9
$_startMicrotaskLoop tear-off trampoline @ main.dart.wasm:0x13b763
$_invokeCallback @ main.dart.wasm:0x1356c8
(anonym) @ main.dart.mjs:661
setTimeout (async)
(anonym) @ sender.ts:94
_send @ sender.ts:91
notifyServiceWorker @ indexed_db.ts:299
(anonym) @ indexed_db.ts:356
await in (anonym) (async)
_withPendingWrite @ indexed_db.ts:330
_remove @ indexed_db.ts:348
removeCurrentUser @ persistence_user_manager.ts:77
directlySetCurrentUser @ auth_impl.ts:790
initializeCurrentUser @ auth_impl.ts:293
await in initializeCurrentUser (async)
(anonym) @ auth_impl.ts:181
Promise.then (async)
queue @ auth_impl.ts:794
_initializeWithPersistence @ auth_impl.ts:159
_initializeAuthInstance @ initialize.ts:86
(anonym) @ register.ts:104
getOrInitializeService @ provider.ts:318
initialize @ provider.ts:242
initializeAuth @ emulator.ts:44
_1659 @ main.dart.mjs:501
$FirebaseAuthWeb.registerWith closure at file:///Users/sma/.pub-cache/hosted/pub.dev/firebase_auth_web-5.11.2/lib/firebase_auth_web.dart:90:32 inner @ main.dart.wasm:0x150152
$FirebaseAuthWeb.registerWith closure at file:///Users/sma/.pub-cache/hosted/pub.dev/firebase_auth_web-5.11.2/lib/firebase_auth_web.dart:90:32 @ main.dart.wasm:0x14ffd8
$closure wrapper at file:///Users/sma/.pub-cache/hosted/pub.dev/firebase_auth_web-5.11.2/lib/firebase_auth_web.dart:90:32 trampoline @ main.dart.wasm:0x150003
$FirebaseCoreWeb.initializeApp closure at file:///Users/sma/.pub-cache/hosted/pub.dev/firebase_core_web-2.15.0/lib/src/firebase_core_web.dart:327:28 @ main.dart.wasm:0x27a5ae
$closure wrapper at file:///Users/sma/.pub-cache/hosted/pub.dev/firebase_core_web-2.15.0/lib/src/firebase_core_web.dart:327:28 trampoline @ main.dart.wasm:0x27ab31
$MappedIterator.moveNext @ main.dart.wasm:0x1435f0
$Future.wait @ main.dart.wasm:0x183843
$FirebaseCoreWeb.initializeApp inner @ main.dart.wasm:0x27aa7b
$_awaitHelperWithTypeCheck closure at org-dartlang-sdk:///dart-sdk/lib/_internal/wasm/lib/async_patch.dart:97:16 @ main.dart.wasm:0x14fd65
$closure wrapper at org-dartlang-sdk:///dart-sdk/lib/_internal/wasm/lib/async_patch.dart:97:16 trampoline @ main.dart.wasm:0x14fdcc
$_RootZone.runUnary @ main.dart.wasm:0x13b242
$_Future._propagateToListeners @ main.dart.wasm:0x13adeb
$_Future._completeWithValue @ main.dart.wasm:0x13aa06
$_Future._asyncCompleteWithValue closure at org-dartlang-sdk:///dart-sdk/lib/async/future_impl.dart:721:29 @ main.dart.wasm:0x13a9b3
$closure wrapper at org-dartlang-sdk:///dart-sdk/lib/async/future_impl.dart:721:29 trampoline @ main.dart.wasm:0x13a9dd
$_startMicrotaskLoop @ main.dart.wasm:0x13b6f9
$_startMicrotaskLoop tear-off trampoline @ main.dart.wasm:0x13b763
$_invokeCallback @ main.dart.wasm:0x1356c8
(anonym) @ main.dart.mjs:661
```

My configuration:
```
$ flutter --version
Flutter 3.22.0-14.0.pre.33 • channel master • https://github.com/flutter/flutter.git
Framework • revision 12d54de962 (2 hours ago) • 2024-04-19 03:10:34 -0400
Engine • revision 25d773ea90
Tools • Dart 3.5.0 (build 3.5.0-76.0.dev) • DevTools 2.35.0-dev.8

$ flutter pub deps
...
firebase_auth 4.19.0
firebase_core 2.30.0
...
```

I think, [this is the code](https://github.com/firebase/flutterfire/blob/2b84feb1b6ec32b1a3605824ed1370b08912184c/packages/firebase_auth/firebase_auth_web/lib/src/interop/auth.dart#L372) where the crash happens.

Here's the line that causes the exception: 
<img width="756" alt="image" src="https://github.com/dart-lang/sdk/assets/14989/3bd5aa9a-c64b-4d68-b787-2755afb659af">

Here's the runtime stack: 
<img width="385" alt="image" src="https://github.com/dart-lang/sdk/assets/14989/a8f6a3e2-1265-412e-be9b-d0b73dc114fa">

I also tried to find the expected type: 
```
    (type $#Closure-0-0 (;15;) (sub $#ClosureBase (struct
      (field $field0 i32)
      (field $field1 (mut i32))
      (field $field2 (ref struct))
      (field $field3 (ref $#Vtable-0-0))
      (field $field4 (ref $_FunctionType)))))
```

I compiled my code with `flutter build web --wasm --no-strip-wasm` and used the `dhttpd` utility as recommended by the documentation to run the code in Chrome 123.0.6312.124.

If I remove all Auth related code, only using Core + Cloudstore, my application works with Wasm. It works as a normal web app in both cases.

**Comment by osa1** [comment]
Is this app open source? Could you share a repro?

**Comment by osa1** [comment]
We've recently landed a fix that may be related to this issue: e44bc22ca3a.

**Comment by sma** [comment]
I check again with 
```
$ flutter --version
Flutter 3.22.0-23.0.pre.18 • channel master • https://github.com/flutter/flutter.git
Framework • revision f4d42c50d3 (2 hours ago) • 2024-05-03 02:43:24 -0400
Engine • revision b8d81a4f6e
Tools • Dart 3.5.0 (build 3.5.0-128.0.dev) • DevTools 2.35.0
```
and it seems the issue has been resolved. Thanks!
